File: rails/app/models/crafting_recipe.rb
class CraftingRecipe < ApplicationRecord
  self.table_name = 'crafting_recipes'

  belongs_to :result_card, class_name: 'Card'
  has_many :required_cards, class_name: 'Card', through: :crafting_ingredients

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:dust_cost_positive, 'Crafting recipe must have a dust cost greater than zero') unless ((dust_cost.nil? || dust_cost > 0))
  end

  def to_s
    dust_cost.to_s
  end

  # Business operations

  def can_craft(player_id)
    # TODO: implement can_craft
    nil
  end

  def execute_craft(player_id)
    # TODO: implement execute_craft
  end

  def disable
    # TODO: implement disable
  end

  def enable
    # TODO: implement enable
  end
end

File: dotnet/CardsProject/Domain/Players/CraftingRecipe.cs
using CardsProject.Domain.Cards;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Players;

public class CraftingRecipe : IValidatableObject
{
    public int Id { get; set; }

    public int DustCost { get; set; } = 0;
    public bool IsAvailable { get; set; } = true;

    public int? ResultCardId { get; set; }
    [ForeignKey(nameof(ResultCardId))]
    public Card? ResultCard { get; set; }

    public ICollection<Card> RequiredCards { get; set; } = new List<Card>();

    // Business operations

    public bool CanCraft(int playerId)
    {
        // TODO: implement can_craft
        return default;
    }

    public void ExecuteCraft(int playerId)
    {
        // TODO: implement execute_craft
    }

    public void Disable()
    {
        // TODO: implement disable
    }

    public void Enable()
    {
        // TODO: implement enable
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( DustCost > 0 ))
            yield return new ValidationResult("Crafting recipe must have a dust cost greater than zero", new[] { nameof(Id) });
    }
}
